internal/pkg/agent/application/upgrade/marker_access_windows.go

```diff
@@ -0,0 +1,66 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+)
+
+const markerAccessTimeout = 10 * time.Second
+const markerAccessBackoffInitialInterval = 50 * time.Millisecond
+const minMarkerAccessRetries = 5
+
+// On Windows, readMarkerFile tries to read the marker file, retrying with
+// randomized exponential backoff up to markerAccessTimeout duration. This retry
+// mechanism is necessary since the marker file could be accessed by multiple
+// processes (the Upgrade Watcher and the main Agent process) at the same time,
+// which could fail on Windows.
+func readMarkerFile(markerFile string) ([]byte, error) {
+	var markerFileBytes []byte
+	readFn := func() (err error) {
+		markerFileBytes, err = os.ReadFile(markerFile)
+	}
+
+	if err := accessMarkerFileWithRetries(readFn); err != nil {
+		return nil, fmt.Errorf("failed to read upgrade marker file [%s] despite retrying: %w", markerFile, err)
+	}
+
+	return markerFileBytes, nil
+}
+
+// On Windows, writeMarkerFile tries to write the marker file, retrying with
+// randomized exponential backoff up to markerAccessTimeout duration. This retry
+// mechanism is necessary since the marker file could be accessed by multiple
+// processes (the Upgrade Watcher and the main Agent process) at the same time,
+// which could fail on Windows.
+func writeMarkerFile(markerFile string, markerBytes []byte) error {
+	writeFn := func() (err error) {
+		return os.WriteFile(markerFile, markerBytes, 0600)
+	}
+
+	if err := accessMarkerFileWithRetries(writeFn); err != nil {
```

Comment:
let's see how it behaves and if there will be any risk of out of order writes. e.g retry writes after another call succeeds. for that we would need to implement some queue mechanism but that sounds like too much and it's possible this will just work